File: java/marsrover/src/main/java/com/birlibirloque/katas/java/marsrover/MarsRover.java
package com.birlibirloque.katas.java.marsrover;

public class MarsRover 
{
	private Position position = null;
	private Direction direction = Direction.North;
	private Grid grid = null;

	public MarsRover(int x, int y, String ad, Grid grid) {
		this.position = new Position(x, y);
		this.direction = Direction.valueOf(ad);
		this.grid = grid;
	}

	public boolean areYouAt (int x, int y, String ad) {
		return (position.areAt(x, y) && (direction.areYouFacing(Direction.valueOf(ad))));
	}

	public void command (String commands) {
		for (int i = 0; i < commands.length(); i++) {
			if (commands.charAt(i) == 'f') {
				this.position.moveStepsByDirection(1,this.direction);
				grid.wrap(this.position);
			}
			else if (commands.charAt(i) == 'b') {
				this.position.moveStepsByDirection(-1,this.direction);
				grid.wrap(this.position);
			}
			else if (commands.charAt(i) == 'r')
				direction = direction.turnRight();
			else if (commands.charAt(i) == 'l')
				direction = direction.turnLeft();
		}
	}
}

File: c++/marsrover/MarsRover.cpp
#include <string.h>
#include "MarsRover.h"

MarsRover::MarsRover () {
	_x = 0;
	_y = 0;
	_direction = 0;
}

MarsRover::MarsRover (int x, int y, char direction) {
	_x = x;
	_y = y;
	for (int i = 0; i < 4; i++) {
		if (direction == direction_char[i]) {
			_direction = i;
		}
	}
}

void MarsRover::command (const char * commands) {
	for (int i = 0; i < strlen(commands); i++) {
		if (commands[i] == 'f')
			_moveByDirection(1);
		else if (commands[i] == 'b')
			_moveByDirection(-1);
		else if (commands[i] == 'r')
			_direction = (_direction + 1) % 4;
		else if (commands[i] == 'l')
			_direction = (_direction + 3) % 4;
	}
}

bool MarsRover::areYouAt (int x, int y, char direction) {
	return ((_x == x) && (_y == y) && (direction_char[_direction] == direction));
}

void MarsRover::_moveByDirection (int sense) {
	if (direction_char[_direction] == 'N') {
		_y += sense;
	}

	else if (direction_char[_direction] == 'S') {
		_y -= sense;
	}

	else if (direction_char[_direction] == 'E') {
		_x += sense;
	}

	else if (direction_char[_direction] == 'W') {
		_x -= sense;
	}
}
